File: crates/workspace-core/src/files.rs
use std::fs;
use std::path::{Path, PathBuf};

use serde::Serialize;

#[derive(Serialize, Debug)]
pub struct DirEntry {
    pub name: String,
    pub path: String,
    pub is_dir: bool,
}

pub fn list_dir(root: &Path, rel: &str) -> Result<Vec<DirEntry>, String> {
    let path = resolve_under_root(root, rel)?;
    let read_dir = fs::read_dir(&path).map_err(|e| e.to_string())?;
    let mut entries = Vec::new();

    for entry in read_dir.flatten() {
        let file_type = entry.file_type().map_err(|e| e.to_string())?;
        let name = entry.file_name().to_string_lossy().into_owned();
        if name.starts_with('.') {
            continue;
        }
        let entry_path = entry.path();
        let rel_path = entry_path
            .strip_prefix(root)
            .unwrap_or(&entry_path)
            .to_string_lossy()
            .into_owned();
        entries.push(DirEntry {
            name,
            path: rel_path,
            is_dir: file_type.is_dir(),
        });
    }

    entries.sort_by(|a, b| b.is_dir.cmp(&a.is_dir).then(a.name.cmp(&b.name)));
    Ok(entries)
}

pub fn read_file(root: &Path, rel: &str) -> Result<String, String> {
    let path = resolve_under_root(root, rel)?;
    fs::read_to_string(path).map_err(|e| e.to_string())
}

pub fn write_file(root: &Path, rel: &str, content: &str) -> Result<(), String> {
    let path = resolve_under_root(root, rel)?;
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    fs::write(path, content).map_err(|e| e.to_string())
}

/// `canonicalize()` requires the full path to already exist, which is
/// fine for `list_dir`/`read_file` but wrong for `write_file` creating a
/// brand-new file: canonicalizing `root.join("untitled.md")` before that
/// file exists always fails with NotFound, so every "create a new file"
/// call was rejected outright before ever reaching `fs::write` (verified
/// directly — `PathBuf::canonicalize()` on a nonexistent path errors,
/// not assumed). Fixed by walking up to whichever prefix of the joined
/// path *does* already exist, canonicalizing and root-checking just that
/// prefix (same escape-prevention guarantee `starts_with` gave before,
/// just checked against an existing ancestor instead of the full leaf),
/// then re-appending the not-yet-existing remainder verbatim.
fn resolve_under_root(root: &Path, rel: &str) -> Result<PathBuf, String> {
    let root = root.canonicalize().unwrap_or_else(|_| root.to_path_buf());
    if rel.is_empty() {
        return Ok(root);
    }
    let joined = root.join(rel);

    let mut existing: &Path = &joined;
    let mut remainder: Vec<std::ffi::OsString> = Vec::new();
    while !existing.exists() {
        let Some(name) = existing.file_name() else {
            return Err("invalid path".into());
        };
        remainder.push(name.to_os_string());
        let Some(parent) = existing.parent() else {
            return Err("invalid path".into());
        };
        existing = parent;
    }

    let mut canonical = existing.canonicalize().map_err(|e| e.to_string())?;
    if !canonical.starts_with(&root) {
        return Err("path escapes workspace root".into());
    }
    for part in remainder.into_iter().rev() {
        canonical.push(part);
    }
    Ok(canonical)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn temp_root() -> PathBuf {
        let dir =
            std::env::temp_dir().join(format!("workspace-core-files-test-{}", std::process::id()));
        let _ = fs::remove_dir_all(&dir);
        fs::create_dir_all(&dir).unwrap();
        dir
    }

    #[test]
    fn write_file_creates_a_brand_new_file() {
        let root = temp_root();
        write_file(&root, "untitled.md", "hello").expect("creating a new file should succeed");
        assert_eq!(read_file(&root, "untitled.md").unwrap(), "hello");
    }

    #[test]
    fn write_file_creates_nested_new_directories_and_file() {
        let root = temp_root();
        write_file(&root, "sub/deeper/note.md", "nested").expect("nested new file should succeed");
        assert_eq!(read_file(&root, "sub/deeper/note.md").unwrap(), "nested");
    }

    #[test]
    fn write_file_still_overwrites_an_existing_file() {
        let root = temp_root();
        write_file(&root, "note.md", "first").unwrap();
        write_file(&root, "note.md", "second").unwrap();
        assert_eq!(read_file(&root, "note.md").unwrap(), "second");
    }

    #[test]
    fn write_file_rejects_escaping_the_root() {
        let root = temp_root();
        let err = write_file(&root, "../../etc/evil.md", "nope").unwrap_err();
        assert!(
            err.contains("escapes"),
            "expected an escape error, got: {err}"
        );
    }

    #[test]
    fn list_dir_sees_a_just_created_file() {
        let root = temp_root();
        write_file(&root, "a.md", "").unwrap();
        let entries = list_dir(&root, "").unwrap();
        assert!(
            entries.iter().any(|e| e.name == "a.md"),
            "expected a.md in {:?}",
            entries
        );
    }
}

File: electron/src/main/files.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { model3dLog } from "./model3d/model3dLog";

// Direct port of crates/workspace-core/src/files.rs.

export interface DirEntry {
  name: string;
  path: string;
  isDir: boolean;
}

/**
 * `fs.realpathSync` requires the full path to already exist — fine for
 * listDir/readFile but wrong for writeFile creating a brand-new file:
 * resolving `root/untitled.md` before that file exists would throw ENOENT,
 * rejecting every "create a new file" call before ever reaching fs.writeFile
 * (ported finding, not re-derived — this exact bug was already found and
 * fixed in the Rust version). Fixed the same way: walk up to whichever
 * prefix of the joined path *does* already exist, resolve just that
 * prefix (same escape-prevention guarantee `startsWith` gave before, just
 * checked against an existing ancestor instead of the full leaf), then
 * re-append the not-yet-existing remainder verbatim.
 */
export function resolveUnderRoot(root: string, rel: string): string {
  const realRoot = fs.existsSync(root) ? fs.realpathSync(root) : root;
  if (!rel) return realRoot;
  const joined = path.join(realRoot, rel);

  let existing = joined;
  const remainder: string[] = [];
  while (!fs.existsSync(existing)) {
    const name = path.basename(existing);
    if (!name || existing === path.dirname(existing)) {
      throw new Error("invalid path");
    }
    remainder.push(name);
    existing = path.dirname(existing);
  }

  let canonical = fs.realpathSync(existing);
  if (!(canonical === realRoot || canonical.startsWith(realRoot + path.sep))) {
    throw new Error("path escapes workspace root");
  }
  for (const part of remainder.reverse()) {
    canonical = path.join(canonical, part);
  }
  return canonical;
}

export function listDir(root: string, rel: string): DirEntry[] {
  const dirPath = resolveUnderRoot(root, rel);
  const realRoot = fs.existsSync(root) ? fs.realpathSync(root) : root;
  const names = fs.readdirSync(dirPath, { withFileTypes: true });
  const entries: DirEntry[] = [];
  for (const entry of names) {
    if (entry.name.startsWith(".")) continue;
    const entryPath = path.join(dirPath, entry.name);
    const relPath = path.relative(realRoot, entryPath);
    entries.push({ name: entry.name, path: relPath, isDir: entry.isDirectory() });
  }
  entries.sort((a, b) => {
    if (a.isDir !== b.isDir) return a.isDir ? -1 : 1;
    return a.name.localeCompare(b.name);
  });
  return entries;
}

export function readFile(root: string, rel: string): string {
  return fs.readFileSync(resolveUnderRoot(root, rel), "utf8");
}

/** null when the file is gone. A tab outlives the file it points at, and
 * the filesystem watcher fires for deletes too, so a missing file is
 * ordinary here rather than a failed IPC call. */
export function readFileIfExists(root: string, rel: string): string | null {
  try {
    return readFile(root, rel);
  } catch (error) {
    if ((error as NodeJS.ErrnoException)?.code === "ENOENT") return null;
    throw error;
  }
}

const BINARY_PREVIEW_MIME_TYPES: Record<string, string> = {
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".webp": "image/webp",
  ".svg": "image/svg+xml",
  ".bmp": "image/bmp",
  ".pdf": "application/pdf",
  ".glb": "model/gltf-binary",
  ".gltf": "model/gltf+json",
  ".obj": "model/obj",
  ".stl": "model/stl",
  ".ply": "application/octet-stream",
  ".dae": "model/vnd.collada+xml",
  ".fbx": "application/octet-stream",
};

export type BinaryFilePreview = { content: string; mimeType: string };

/** Base64 file content for the File Viewer pane (Orca parity —
 * useLocalImageSrc.ts's readImagePreview: read via IPC and hand the
 * renderer bytes to build its own blob: URL, rather than a file:// URL —
 * Chromium blocks file:// resource loads from a page not itself loaded via
 * file://, which a dev-mode Vite server never is). Goes through the same
 * resolveUnderRoot confinement as every other file op here. */
export function readFileBinaryPreview(root: string, rel: string): BinaryFilePreview | null {
  const target = resolveUnderRoot(root, rel);
  const ext = path.extname(target).toLowerCase();
  const mimeType = BINARY_PREVIEW_MIME_TYPES[ext];
  if (!mimeType) {
    model3dLog("binary_preview_unsupported_ext", {
      source: "main",
      relativePath: rel,
      absolutePath: target,
      ext,
    });
    return null;
  }
  try {
    const content = fs.readFileSync(target).toString("base64");
    const isModel =
      ext === ".glb" ||
      ext === ".gltf" ||
      ext === ".obj" ||
      ext === ".stl" ||
      ext === ".ply" ||
      ext === ".dae" ||
      ext === ".fbx";
    if (isModel) {
      model3dLog("binary_preview_read_ok", {
        source: "main",
        relativePath: rel,
        absolutePath: target,
        ext,
        mimeType,
        byteLength: Buffer.byteLength(content, "base64"),
      });
    }
    return { content, mimeType };
  } catch (err) {
    model3dLog("binary_preview_read_failed", {
      source: "main",
      relativePath: rel,
      absolutePath: target,
      ext,
      mimeType,
      error: err instanceof Error ? err.message : String(err),
    });
    return null;
  }
}

export function writeFile(root: string, rel: string, content: string): void {
  const target = resolveUnderRoot(root, rel);
  fs.mkdirSync(path.dirname(target), { recursive: true });
  fs.writeFileSync(target, content, "utf8");
}

export function createDir(root: string, rel: string): void {
  fs.mkdirSync(resolveUnderRoot(root, rel), { recursive: true });
}

/** `rel` must already exist — checked explicitly rather than forwarding a
 * confusing OS error. */
export function deletePath(root: string, rel: string): void {
  const target = resolveUnderRoot(root, rel);
  const stat = fs.lstatSync(target);
  if (stat.isDirectory()) {
    fs.rmSync(target, { recursive: true, force: true });
  } else {
    fs.unlinkSync(target);
  }
}

export function renamePath(root: string, fromRel: string, toRel: string): void {
  const from = resolveUnderRoot(root, fromRel);
  const to = resolveUnderRoot(root, toRel);
  fs.mkdirSync(path.dirname(to), { recursive: true });
  fs.renameSync(from, to);
}
